File: backend/src/lib/rate-limit.ts
// Abuse-protection rate limits + admin-configured quotas.
//
// Strictly NON-MONETARY — lintel has no billing. Two layers:
//
//   1. Rate limits (always on, env-tunable): per-member open cooldown,
//      per-member opens/hour, per-account opens/hour ceiling, and a
//      per-sender chat-message throttle for webhook floods.
//   2. Quotas (admin policy, off by default): per-location optional caps in
//      location_settings — opens per member per UTC day, opens per location
//      per UTC day. NULL = unlimited.
//
// Counting uses fixed-window counters in Postgres (rate_limit_counters),
// updated atomically via the SECURITY DEFINER app.rate_limit_* functions
// (see migrations/20260505010000_rate_limits.sql). The table itself is
// internal: FORCEd RLS with no policies, so tenants can neither inspect nor
// exhaust each other's counters.
//
// Enforcement is centralized in logAccess() (routes/access.ts), the single
// choke point shared by portal, API, WhatsApp and Slack opens — no open path
// can bypass it.
//
// POLICY CHOICES (documented on purpose):
//   - Quotas apply to every open path (portal, chat, API) for consistency,
//     EXCEPT account owners/admins, who are exempt from quotas — but NOT
//     from rate limits. An admin can always let the plumber in, but a
//     runaway admin script still hits the cooldown/hourly caps.
//   - Admin opens still increment the location's daily counter (they are
//     real gate movements), they just cannot be denied by a quota.
//   - 'close' commands are never limited: closing a gate is the safe
//     direction and should never be refused.
//   - Denied attempts do NOT consume counters — counters track successful
//     opens only, so they double as honest usage numbers for the UI.
//     Enforcement is EXACT under concurrency: every limit is consumed via an
//     atomic increment-if-under-cap (app.rate_limit_try_bump) and the
//     cooldown via an atomic sentinel claim (app.rate_limit_claim_cooldown),
//     so N simultaneous opens can never land a counter past its cap.
//   - FAILURE MODE: if the counter store errors, the open is ALLOWED and the
//     success audit row is tagged error='rate_limit_check_failed'. A gate is
//     physical access — availability wins for enforcement, visibility is
//     preserved through the audit tag. (Fail-open reviewed 2026-07-17:
//     accepted — physical access availability wins; audit-tagged.)

import type { TxSql } from './db.ts';
import { getEnv } from './env.ts';

// ---------------------------------------------------------------------------
// Config
// ---------------------------------------------------------------------------

export type RateLimitConfig = {
  /** Min seconds between successful opens per (member, access point). 0 disables. */
  openCooldownS: number;
  /** Max successful opens per member per hour. */
  opensPerHour: number;
  /** Max inbound chat messages per sender per minute before the bot goes quiet. */
  chatMsgsPerMin: number;
  /** Max successful opens per account per hour (runaway-integration ceiling). */
  accountOpensPerHour: number;
};

export const RATE_LIMIT_DEFAULTS: RateLimitConfig = {
  openCooldownS: 10,
  opensPerHour: 30,
  chatMsgsPerMin: 10,
  accountOpensPerHour: 500,
};

/**
 * Parse one env value. Accepts non-negative integers; 0 is a valid explicit
 * value (0 cooldown = disabled; 0 cap = block everything, a kill switch).
 * Missing / malformed / negative values fall back to the default.
 */
export function parseRateLimitValue(raw: string | undefined, fallback: number): number {
  if (raw === undefined || raw.trim() === '') return fallback;
  const n = Number(raw.trim());
  if (!Number.isInteger(n) || n < 0) return fallback;
  return n;
}

export function parseRateLimitConfig(raw: {
  RATE_OPEN_COOLDOWN_S?: string;
  RATE_OPENS_PER_HOUR?: string;
  RATE_CHAT_MSGS_PER_MIN?: string;
  RATE_ACCOUNT_OPENS_PER_HOUR?: string;
}): RateLimitConfig {
  return {
    openCooldownS: parseRateLimitValue(raw.RATE_OPEN_COOLDOWN_S, RATE_LIMIT_DEFAULTS.openCooldownS),
    opensPerHour: parseRateLimitValue(raw.RATE_OPENS_PER_HOUR, RATE_LIMIT_DEFAULTS.opensPerHour),
    chatMsgsPerMin: parseRateLimitValue(
      raw.RATE_CHAT_MSGS_PER_MIN,
      RATE_LIMIT_DEFAULTS.chatMsgsPerMin,
    ),
    accountOpensPerHour: parseRateLimitValue(
      raw.RATE_ACCOUNT_OPENS_PER_HOUR,
      RATE_LIMIT_DEFAULTS.accountOpensPerHour,
    ),
  };
}

export function getRateLimitConfig(): RateLimitConfig {
  return parseRateLimitConfig(getEnv());
}

// ---------------------------------------------------------------------------
// Runtime overrides (instance_settings, set by the instance admin)
// ---------------------------------------------------------------------------
// The gateway operator can override the env-configured limits at runtime via
// PATCH /admin/limits. Overrides persist in the internal instance_settings
// table under key 'rate_limits' as a PARTIAL object, e.g. {"opens_per_hour":1}.
// Resolution order per field: db override > env var > built-in default.

/** External (snake_case) override field → internal config key. */
export const RATE_LIMIT_OVERRIDE_FIELDS = {
  open_cooldown_s: 'openCooldownS',
  opens_per_hour: 'opensPerHour',
  chat_msgs_per_min: 'chatMsgsPerMin',
  account_opens_per_hour: 'accountOpensPerHour',
} as const;

export type RateLimitOverrideField = keyof typeof RATE_LIMIT_OVERRIDE_FIELDS;

/** Partial set of overridden values, keyed by external field name. */
export type RateLimitOverrides = Partial<Record<RateLimitOverrideField, number>>;

export const INSTANCE_RATE_LIMITS_KEY = 'rate_limits';

/**
 * Parse the stored jsonb override object defensively: only non-negative
 * integers are accepted; anything else (missing key, garbage, negatives,
 * floats) is ignored so a corrupted settings row can never wedge the limiter.
 */
export function parseStoredOverrides(raw: unknown): RateLimitOverrides {
  const out: RateLimitOverrides = {};
  if (raw === null || typeof raw !== 'object' || Array.isArray(raw)) return out;
  const obj = raw as Record<string, unknown>;
  for (const field of Object.keys(RATE_LIMIT_OVERRIDE_FIELDS) as RateLimitOverrideField[]) {
    const v = obj[field];
    if (typeof v === 'number' && Number.isInteger(v) && v >= 0) out[field] = v;
  }
  return out;
}

/** Apply overrides on top of the env/default config. db > env > default. */
export function mergeRateLimitConfig(
  base: RateLimitConfig,
  overrides: RateLimitOverrides,
): RateLimitConfig {
  const cfg = { ...base };
  for (const [field, key] of Object.entries(RATE_LIMIT_OVERRIDE_FIELDS) as [
    RateLimitOverrideField,
    keyof RateLimitConfig,
  ][]) {
    const v = overrides[field];
    if (v !== undefined) cfg[key] = v;
  }
  return cfg;
}

/** Read the persisted override object (empty when none / unreadable). */
export async function readRateLimitOverrides(tx: TxSql): Promise<RateLimitOverrides> {
  const rows = await tx<{ v: unknown }[]>`
    select app.instance_setting_get(${INSTANCE_RATE_LIMITS_KEY}) as v
  `;
  return parseStoredOverrides(rows[0]?.v ?? null);
}

/**
 * Effective config for enforcement: db override > env > default.
 * FAILURE MODE: if the settings read errors (e.g. migration not applied),
 * fall back to the env config — never block on the settings store. The read
 * runs in a savepoint so a failure can't poison the caller's transaction.
 */
export async function resolveRateLimitConfig(tx: TxSql): Promise<RateLimitConfig> {
  const envCfg = getRateLimitConfig();
  try {
    const overrides = await tx.savepoint(async (stx) => await readRateLimitOverrides(stx));
    return mergeRateLimitConfig(envCfg, overrides);
  } catch (err) {
    console.error('rate_limit_overrides_read_failed', err);
    return envCfg;
  }
}

// ---------------------------------------------------------------------------
// Window math (pure, unit-tested)
// ---------------------------------------------------------------------------

export const HOUR_S = 3600;
export const MINUTE_S = 60;
export const DAY_S = 86400;

/** UTC epoch-aligned fixed-window start containing `now`. */
export function fixedWindowStart(now: Date, windowS: number): Date {
  const epochS = Math.floor(now.getTime() / 1000);
  return new Date(Math.floor(epochS / windowS) * windowS * 1000);
}

/** Whole seconds until the current fixed window rolls over (min 1). */
export function secondsUntilWindowEnd(now: Date, windowS: number): number {
  const start = fixedWindowStart(now, windowS).getTime();
  const end = start + windowS * 1000;
  return Math.max(1, Math.ceil((end - now.getTime()) / 1000));
}

/** Remaining cooldown seconds; 0 when the cooldown has fully elapsed. */
export function cooldownRemainingS(lastOpenAt: Date, now: Date, cooldownS: number): number {
  const elapsedS = (now.getTime() - lastOpenAt.getTime()) / 1000;
  if (elapsedS >= cooldownS) return 0;
  return Math.max(1, Math.ceil(cooldownS - elapsedS));
}

// ---------------------------------------------------------------------------
// Open-path enforcement
// ---------------------------------------------------------------------------

export type OpenLimitInput = {
  /** Authenticated member opening the gate, when known. */
  userId: string | null;
  /** E.164 phone of the opener (WhatsApp path — members and visitor grants). */
  phoneE164?: string | null;
  accessPointId: string;
  locationId: string;
  accountId: string;
  now?: Date;
};

export type OpenLimitDecision =
  | {
      allowed: true;
      /** Set when the counter store failed and enforcement was skipped. */
      degraded?: boolean;
    }
  | {
      allowed: false;
      reason: 'rate_limited' | 'quota_exceeded';
      /** Seconds after which a retry can succeed. */
      retryAfterS: number;
      /** Which limit tripped (for logs/tests, not user-facing). */
      limit:
        | 'open_cooldown'
        | 'member_opens_per_hour'
        | 'account_opens_per_hour'
        | 'member_opens_per_day'
        | 'location_opens_per_day';
    };

function openSubject(input: OpenLimitInput): string | null {
  if (input.userId) return `user:${input.userId}`;
  if (input.phoneE164) return `phone:${input.phoneE164}`;
  return null;
}

async function bump(
  tx: TxSql,
  scope: string,
  subject: string,
  windowStart: Date,
  amount = 1,
): Promise<number> {
  const rows = await tx<{ count: number }[]>`
    select app.rate_limit_bump(${scope}, ${subject}, ${windowStart}, ${amount}) as count
  `;
  return Number(rows[0]?.count ?? 0);
}

/**
 * Atomically consume one unit of a capped counter. Returns the
 * post-increment count when consumed, or null when the cap would be
 * exceeded — in which case NOTHING was consumed. cap === null means
 * uncapped (plain increment).
 */
async function tryBump(
  tx: TxSql,
  scope: string,
  subject: string,
  windowStart: Date,
  cap: number | null,
): Promise<number | null> {
  const rows = await tx<{ count: number | null }[]>`
    select app.rate_limit_try_bump(${scope}, ${subject}, ${windowStart}, ${cap}) as count
  `;
  const v = rows[0]?.count;
  return v === null || v === undefined ? null : Number(v);
}

/**
 * Atomically claim the cooldown sentinel: true iff the previous successful
 * open is at least cooldownS old (the sentinel is then refreshed). A failed
 * claim leaves the sentinel untouched. cooldownS <= 0 always claims.
 */
async function claimCooldown(tx: TxSql, subject: string, cooldownS: number): Promise<boolean> {
  const rows = await tx<{ ok: boolean }[]>`
    select app.rate_limit_claim_cooldown(${subject}, ${cooldownS}) as ok
  `;
  return rows[0]?.ok === true;
}

/**
 * Check every open limit and, when allowed, consume the counters.
 *
 * Increment-then-check, one atomic statement per limit: each counter is
 * consumed via app.rate_limit_try_bump, which increments ONLY when the
 * post-increment count stays within the cap (ON CONFLICT DO UPDATE ...
 * WHERE count < cap). Concurrent opens serialize on the counter row, so the
 * bound is EXACT — N simultaneous requests at count = cap-1 admit exactly
 * one, never cap+N. The cooldown is claimed the same way (atomic
 * conditional UPDATE on the sentinel row), so two simultaneous opens cannot
 * both pass it. When a later limit denies the attempt, counters already
 * consumed for THIS attempt are handed back (bump -1), preserving the
 * documented semantics: denials never consume, and counters exactly equal
 * successful opens.
 *
 * Throws on counter-store failure — use guardedCheckOpenLimits() for the
 * fail-open wrapper.
 */
export async function checkAndConsumeOpenLimits(
  tx: TxSql,
  input: OpenLimitInput,
): Promise<OpenLimitDecision> {
  // db override > env > default (see resolveRateLimitConfig).
  const cfg = await resolveRateLimitConfig(tx);
  const now = input.now ?? new Date();
  const subject = openSubject(input);
  const hourStart = fixedWindowStart(now, HOUR_S);
  const dayStart = fixedWindowStart(now, DAY_S);
  const acctSubject = `acct:${input.accountId}`;
  const locSubject = `loc:${input.locationId}`;

  // Quota config + admin exemption in one round trip. Runs under the
  // caller's RLS context: members can see their own account_members rows and
  // the location settings; the anon webhook context passes the NULL-user
  // policies.
  const cfgRows = await tx<
    { member_quota: number | null; location_quota: number | null; is_admin: boolean }[]
  >`
    select
      ls.max_opens_per_member_per_day as member_quota,
      ls.max_opens_per_location_per_day as location_quota,
      (
        ${input.userId ?? null}::uuid is not null
        and exists (
          select 1 from account_members am
          where am.account_id = ${input.accountId}::uuid
            and am.user_id = ${input.userId ?? null}::uuid
            and am.status = 'active'
            and am.role in ('owner', 'admin')
        )
      ) as is_admin
    from (select 1) as one
    left join location_settings ls on ls.location_id = ${input.locationId}::uuid
  `;
  const quota = cfgRows[0] ?? { member_quota: null, location_quota: null, is_admin: false };

  const cdSubject = subject ? `${subject}|ap:${input.accessPointId}` : null;

  // Counters consumed for THIS attempt so far — handed back if a later
  // limit denies it (denials must never consume).
  const consumed: Array<{ scope: string; subject: string; windowStart: Date }> = [];
  const handBack = async (): Promise<void> => {
    for (const c of consumed) await bump(tx, c.scope, c.subject, c.windowStart, -1);
  };

  // --- Cooldown fast-path (read-only) --------------------------------------
  // Cheap early deny with a precise Retry-After. The AUTHORITATIVE check is
  // the atomic sentinel claim at the end — this read alone would race.

  if (cdSubject && cfg.openCooldownS > 0) {
    const rows = await tx<{ last: Date | null }[]>`
      select app.rate_limit_last('open_cd', ${cdSubject}) as last
    `;
    const last = rows[0]?.last ? new Date(rows[0].last) : null;
    if (last) {
      const remaining = cooldownRemainingS(last, now, cfg.openCooldownS);
      if (remaining > 0) {
        return { allowed: false, reason: 'rate_limited', retryAfterS: remaining, limit: 'open_cooldown' };
      }
    }
  }

  // --- Rate limits (everyone, admins included) -----------------------------

  if (subject) {
    const memberHour = await tryBump(tx, 'opens_1h', subject, hourStart, cfg.opensPerHour);
    if (memberHour === null) {
      await handBack();
      return {
        allowed: false,
        reason: 'rate_limited',
        retryAfterS: secondsUntilWindowEnd(now, HOUR_S),
        limit: 'member_opens_per_hour',
      };
    }
    consumed.push({ scope: 'opens_1h', subject, windowStart: hourStart });
  }

  const acctHour = await tryBump(tx, 'acct_opens_1h', acctSubject, hourStart, cfg.accountOpensPerHour);
  if (acctHour === null) {
    await handBack();
    return {
      allowed: false,
      reason: 'rate_limited',
      retryAfterS: secondsUntilWindowEnd(now, HOUR_S),
      limit: 'account_opens_per_hour',
    };
  }
  consumed.push({ scope: 'acct_opens_1h', subject: acctSubject, windowStart: hourStart });

  // --- Quotas (admin policy; owners/admins exempt) -------------------------
  // Admin-exempt opens still increment the daily counters (they are real
  // gate movements) — cap null = uncapped increment.

  if (subject) {
    const mdSubject = `${subject}|loc:${input.locationId}`;
    const memberDayCap =
      !quota.is_admin && quota.member_quota !== null ? Number(quota.member_quota) : null;
    const memberDay = await tryBump(tx, 'opens_1d', mdSubject, dayStart, memberDayCap);
    if (memberDay === null) {
      await handBack();
      return {
        allowed: false,
        reason: 'quota_exceeded',
        retryAfterS: secondsUntilWindowEnd(now, DAY_S),
        limit: 'member_opens_per_day',
      };
    }
    consumed.push({ scope: 'opens_1d', subject: mdSubject, windowStart: dayStart });
  }

  const locDayCap =
    !quota.is_admin && quota.location_quota !== null ? Number(quota.location_quota) : null;
  const locDay = await tryBump(tx, 'loc_opens_1d', locSubject, dayStart, locDayCap);
  if (locDay === null) {
    await handBack();
    return {
      allowed: false,
      reason: 'quota_exceeded',
      retryAfterS: secondsUntilWindowEnd(now, DAY_S),
      limit: 'location_opens_per_day',
    };
  }
  consumed.push({ scope: 'loc_opens_1d', subject: locSubject, windowStart: dayStart });

  // --- Cooldown claim (authoritative, atomic) ------------------------------
  // Last on purpose: the sentinel's updated_at is only refreshed when the
  // open is fully allowed, so a denied attempt never restarts anyone's
  // cooldown. With cooldown disabled (0) the claim always succeeds and
  // still records the last successful open.

  if (cdSubject) {
    const claimed = await claimCooldown(tx, cdSubject, cfg.openCooldownS);
    if (!claimed) {
      await handBack();
      // Lost a concurrent race for the sentinel — recompute Retry-After
      // from the winner's refreshed timestamp.
      const rows = await tx<{ last: Date | null }[]>`
        select app.rate_limit_last('open_cd', ${cdSubject}) as last
      `;
      const last = rows[0]?.last ? new Date(rows[0].last) : null;
      const remaining = last ? cooldownRemainingS(last, now, cfg.openCooldownS) : cfg.openCooldownS;
      return {
        allowed: false,
        reason: 'rate_limited',
        retryAfterS: Math.max(1, remaining),
        limit: 'open_cooldown',
      };
    }
  }

  return { allowed: true };
}

/**
 * Fail-open wrapper. If the counter store errors, roll back to a savepoint
 * (Postgres poisons the surrounding transaction after any error, so the
 * caller's audit-log insert would otherwise fail too) and ALLOW the open,
 * flagged degraded so logAccess records error='rate_limit_check_failed'.
 * A gate is physical access — availability wins for enforcement,
 * visibility is preserved. (Fail-open reviewed 2026-07-17: accepted —
 * physical access availability wins; audit-tagged.)
 */
export async function guardedCheckOpenLimits(
  tx: TxSql,
  input: OpenLimitInput,
): Promise<OpenLimitDecision> {
  try {
    return await tx.savepoint(async (stx) => await checkAndConsumeOpenLimits(stx, input));
  } catch (err) {
    console.error('rate_limit_check_failed', err);
    return { allowed: true, degraded: true };
  }
}

// ---------------------------------------------------------------------------
// Phone-OTP flow limits
// ---------------------------------------------------------------------------
// Persistent fixed-window counters for the phone verification flow
// (src/routes/phones.ts). These exist because the per-challenge attempt cap
// alone is resettable: POST /me/phones re-add restarts the challenge with a
// fresh code and attempts = 0, so without a persistent counter an attacker
// could grind the 10^6 code space indefinitely.
//
//   otp_start  — challenges started per USER per hour (also bounds how many
//                fresh codes / WhatsApp sends one user can trigger).
//   otp_verify — verify attempts per PHONE ROW per hour, surviving challenge
//                restarts within the window.
//
// FAILURE MODE: unlike the gate-open path these are NOT fail-open — an OTP
// verification is an identity claim, not physical access, so a counter-store
// error propagates (request fails closed).

export const OTP_START_MAX_PER_HOUR = 5;
export const OTP_VERIFY_MAX_PER_HOUR = 15;

export type OtpLimitDecision = { allowed: true } | { allowed: false; retryAfterS: number };

async function consumeOtpLimit(
  tx: TxSql,
  scope: 'otp_start' | 'otp_verify',
  subject: string,
  cap: number,
  now: Date,
): Promise<OtpLimitDecision> {
  const hourStart = fixedWindowStart(now, HOUR_S);
  const count = await tryBump(tx, scope, subject, hourStart, cap);
  if (count === null) {
    return { allowed: false, retryAfterS: secondsUntilWindowEnd(now, HOUR_S) };
  }
  return { allowed: true };
}

/** Consume one OTP-challenge start for a user; deny past the hourly cap. */
export async function consumeOtpStartLimit(
  tx: TxSql,
  userId: string,
  now: Date = new Date(),
): Promise<OtpLimitDecision> {
  return await consumeOtpLimit(tx, 'otp_start', `user:${userId}`, OTP_START_MAX_PER_HOUR, now);
}

/** Consume one OTP verify attempt for a phone row; deny past the hourly cap. */
export async function consumeOtpVerifyLimit(
  tx: TxSql,
  phoneRowId: string,
  now: Date = new Date(),
): Promise<OtpLimitDecision> {
  return await consumeOtpLimit(
    tx,
    'otp_verify',
    `phone_row:${phoneRowId}`,
    OTP_VERIFY_MAX_PER_HOUR,
    now,
  );
}

// ---------------------------------------------------------------------------
// Chat flood throttle
// ---------------------------------------------------------------------------

/**
 * Count an inbound chat message for a sender subject ('phone:+27...',
 * 'slack:U123', 'tg:12345'). Returns quiet=true when the sender exceeded
 * RATE_CHAT_MSGS_PER_MIN in the current minute window — the bot should stop
 * replying (but the webhook must still return 200 so the provider does not
 * amplify with retries). Fail-open on counter-store errors: keep replying.
 */
export async function noteChatMessage(
  tx: TxSql,
  subject: string,
  now: Date = new Date(),
): Promise<{ quiet: boolean }> {
  const cfg = await resolveRateLimitConfig(tx);
  try {
    const count = await tx.savepoint(
      async (stx) => await bump(stx, 'chat_1m', subject, fixedWindowStart(now, MINUTE_S)),
    );
    return { quiet: count > cfg.chatMsgsPerMin };
  } catch (err) {
    console.error('chat_rate_limit_check_failed', err);
    return { quiet: false };
  }
}

// ---------------------------------------------------------------------------
// User-facing denial copy (shared by WhatsApp + Slack)
// ---------------------------------------------------------------------------

export function chatDenialMessage(denial: {
  reason: 'rate_limited' | 'quota_exceeded' | 'account_suspended' | 'user_disabled';
  retry_after_s: number;
}): string {
  if (denial.reason === 'account_suspended') {
    return 'This account has been suspended by the gateway operator — the gate cannot be opened. Contact your operator for help.';
  }
  if (denial.reason === 'user_disabled') {
    return 'Your lintel user has been disabled by the gateway operator — the gate cannot be opened. Contact your operator for help.';
  }
  if (denial.reason === 'quota_exceeded') {
    const base = getEnv().APP_PUBLIC_URL.replace(/\/$/, '');
    return `Daily limit reached for this location — contact your admin. The web portal: ${base}/app`;
  }
  const mins = Math.max(1, Math.ceil(denial.retry_after_s / 60));
  return `Too many opens — try again in ~${mins} min.`;
}

File: gateway/internal/store/ratelimit.go
package store

// Abuse-protection rate limits + admin quotas, porting backend
// src/lib/rate-limit.ts onto SQLite. Strictly NON-MONETARY.
//
// Counting uses fixed-window counters (rate_limit_counters) plus a cooldown
// sentinel table (rate_limit_cooldowns), updated with atomic
// increment-if-under-cap statements. SQLite has one writer and the store
// runs a single serialized connection, so each statement is atomic and the
// bounds are EXACT under concurrency (the Postgres backend needed
// SECURITY DEFINER upsert functions for the same guarantee).
//
// POLICY (ported verbatim):
//   - Quotas exempt account owners/admins; rate limits never exempt anyone.
//   - Admin opens still increment daily counters (real gate movements).
//   - 'close' is never limited.
//   - Denials never consume: counters equal successful opens exactly.
//     Counters consumed for a denied attempt are handed back (bump -1).
//   - FAILURE MODE: counter-store errors ALLOW the open, tagged
//     error='rate_limit_check_failed' in the audit row (physical access —
//     availability wins; reviewed + accepted upstream 2026-07-17).

import (
	"bytes"
	"context"
	"database/sql"
	"encoding/json"
	"strconv"
	"strings"
)

// Window sizes (seconds).
const (
	MinuteS int64 = 60
	HourS   int64 = 3600
	DayS    int64 = 86400
)

// FixedWindowStart is the UTC epoch-aligned window containing now.
func FixedWindowStart(nowUnix, windowS int64) int64 {
	return nowUnix - nowUnix%windowS
}

// SecondsUntilWindowEnd is the whole seconds until rollover (min 1).
func SecondsUntilWindowEnd(nowUnix, windowS int64) int64 {
	rem := FixedWindowStart(nowUnix, windowS) + windowS - nowUnix
	if rem < 1 {
		return 1
	}
	return rem
}

// CooldownRemainingS is the remaining cooldown (0 when elapsed, min 1 otherwise).
func CooldownRemainingS(lastOpenAt, nowUnix, cooldownS int64) int64 {
	elapsed := nowUnix - lastOpenAt
	if elapsed >= cooldownS {
		return 0
	}
	rem := cooldownS - elapsed
	if rem < 1 {
		return 1
	}
	return rem
}

// ---------------------------------------------------------------------------
// Config: built-in defaults < env < instance_settings overrides
// ---------------------------------------------------------------------------

// RateLimitConfig mirrors the backend's RateLimitConfig.
type RateLimitConfig struct {
	OpenCooldownS       int64 // min seconds between opens per (subject, access point); 0 disables
	OpensPerHour        int64 // per member
	ChatMsgsPerMin      int64 // webhook flood throttle (channel seam)
	AccountOpensPerHour int64 // runaway-integration ceiling
}

// RateLimitDefaults are the built-in values (backend RATE_LIMIT_DEFAULTS).
var RateLimitDefaults = RateLimitConfig{
	OpenCooldownS:       10,
	OpensPerHour:        30,
	ChatMsgsPerMin:      10,
	AccountOpensPerHour: 500,
}

// ParseRateLimitValue accepts non-negative integers; 0 is valid (0 cooldown =
// disabled; 0 cap = kill switch). Missing/garbage/negative → fallback.
func ParseRateLimitValue(raw string, fallback int64) int64 {
	raw = strings.TrimSpace(raw)
	if raw == "" {
		return fallback
	}
	n, err := strconv.ParseInt(raw, 10, 64)
	if err != nil || n < 0 {
		return fallback
	}
	return n
}

// ParseRateLimitConfig builds the env layer from a lookup func (os.Getenv in
// production, a map in tests). Env vars are the backend's names.
func ParseRateLimitConfig(getenv func(string) string) RateLimitConfig {
	d := RateLimitDefaults
	return RateLimitConfig{
		OpenCooldownS:       ParseRateLimitValue(getenv("RATE_OPEN_COOLDOWN_S"), d.OpenCooldownS),
		OpensPerHour:        ParseRateLimitValue(getenv("RATE_OPENS_PER_HOUR"), d.OpensPerHour),
		ChatMsgsPerMin:      ParseRateLimitValue(getenv("RATE_CHAT_MSGS_PER_MIN"), d.ChatMsgsPerMin),
		AccountOpensPerHour: ParseRateLimitValue(getenv("RATE_ACCOUNT_OPENS_PER_HOUR"), d.AccountOpensPerHour),
	}
}

// InstanceRateLimitsKey is the instance_settings key for runtime overrides.
const InstanceRateLimitsKey = "rate_limits"

// RateLimitOverrides is the partial override object stored by PATCH
// /admin/limits, keyed by the external snake_case field names.
type RateLimitOverrides map[string]int64

// RateLimitOverrideFields are the accepted external field names.
var RateLimitOverrideFields = []string{
	"open_cooldown_s", "opens_per_hour", "chat_msgs_per_min", "account_opens_per_hour",
}

// ParseStoredOverrides parses the stored json defensively: only non-negative
// integers on known fields are accepted — a corrupted settings row can never
// wedge the limiter.
func ParseStoredOverrides(raw json.RawMessage) RateLimitOverrides {
	out := RateLimitOverrides{}
	if len(raw) == 0 {
		return out
	}
	dec := json.NewDecoder(bytes.NewReader(raw))
	dec.UseNumber()
	var m map[string]any
	if err := dec.Decode(&m); err != nil {
		return out
	}
	for _, f := range RateLimitOverrideFields {
		num, ok := m[f].(json.Number)
		if !ok {
			continue
		}
		n, err := num.Int64()
		// Reject floats (Int64 errors) and negatives; strconv round-trip
		// guards "1.0"-style numbers that Int64 would accept on some paths.
		if err != nil || n < 0 || num.String() != strconv.FormatInt(n, 10) {
			continue
		}
		out[f] = n
	}
	return out
}

// MergeRateLimitConfig applies overrides on top of the env/default config.
func MergeRateLimitConfig(base RateLimitConfig, o RateLimitOverrides) RateLimitConfig {
	cfg := base
	if v, ok := o["open_cooldown_s"]; ok {
		cfg.OpenCooldownS = v
	}
	if v, ok := o["opens_per_hour"]; ok {
		cfg.OpensPerHour = v
	}
	if v, ok := o["chat_msgs_per_min"]; ok {
		cfg.ChatMsgsPerMin = v
	}
	if v, ok := o["account_opens_per_hour"]; ok {
		cfg.AccountOpensPerHour = v
	}
	return cfg
}

// ReadRateLimitOverrides reads the persisted override object (empty when
// none / unreadable).
func (s *Store) ReadRateLimitOverrides(ctx context.Context) RateLimitOverrides {
	raw, err := s.InstanceSettingGet(ctx, InstanceRateLimitsKey)
	if err != nil {
		return RateLimitOverrides{}
	}
	return ParseStoredOverrides(raw)
}

// ResolveRateLimitConfig is the effective enforcement config: db override >
// env > default. FAILURE MODE: an unreadable settings row falls back to the
// env config — the limiter never blocks on the settings store.
func (s *Store) ResolveRateLimitConfig(ctx context.Context, envCfg RateLimitConfig) RateLimitConfig {
	return MergeRateLimitConfig(envCfg, s.ReadRateLimitOverrides(ctx))
}

// ---------------------------------------------------------------------------
// Counter primitives (each one statement = atomic under SQLite's writer lock)
// ---------------------------------------------------------------------------

// bump adjusts a counter unconditionally (used with -1 to hand back a unit
// consumed by an attempt that a later limit denied). Floors at 0.
func (s *Store) rateLimitBump(ctx context.Context, scope, subject string, windowStart, amount int64) error {
	if _, err := s.db.ExecContext(ctx,
		`INSERT INTO rate_limit_counters (scope, subject, window_start, count)
		 VALUES (?, ?, ?, 0) ON CONFLICT (scope, subject, window_start) DO NOTHING`,
		scope, subject, windowStart); err != nil {
		return err
	}
	// The adjustment itself is one atomic UPDATE; interleavings with other
	// bumps still net correctly.
	_, err := s.db.ExecContext(ctx,
		`UPDATE rate_limit_counters SET count = max(count + ?, 0)
		 WHERE scope = ? AND subject = ? AND window_start = ?`,
		amount, scope, subject, windowStart)
	return err
}

// rateLimitTryBump atomically consumes one unit of a capped counter.
// Returns consumed=false (nothing consumed) when the cap would be exceeded.
// cap == nil means uncapped; cap == 0 denies everything (kill switch).
func (s *Store) rateLimitTryBump(ctx context.Context, scope, subject string, windowStart int64, cap *int64) (bool, error) {
	if cap == nil {
		return true, s.rateLimitBump(ctx, scope, subject, windowStart, 1)
	}
	if *cap <= 0 {
		return false, nil
	}
	res, err := s.db.ExecContext(ctx,
		`INSERT INTO rate_limit_counters (scope, subject, window_start, count)
		 VALUES (?, ?, ?, 1)
		 ON CONFLICT (scope, subject, window_start) DO UPDATE SET count = count + 1
		 WHERE count < ?`,
		scope, subject, windowStart, *cap)
	if err != nil {
		return false, err
	}
	n, err := res.RowsAffected()
	if err != nil {
		return false, err
	}
	return n > 0, nil
}

// rateLimitLast reads the cooldown sentinel's last successful open.
func (s *Store) rateLimitLast(ctx context.Context, subject string) (int64, bool, error) {
	var last int64
	err := s.db.QueryRowContext(ctx,
		`SELECT last_open_at FROM rate_limit_cooldowns WHERE subject = ?`, subject).Scan(&last)
	if err == sql.ErrNoRows {
		return 0, false, nil
	}
	if err != nil {
		return 0, false, err
	}
	return last, true, nil
}

// rateLimitClaimCooldown atomically claims the cooldown sentinel: true iff
// the previous successful open is at least cooldownS old (sentinel then
// refreshed). A failed claim leaves the sentinel untouched. cooldownS <= 0
// always claims (and still records the last successful open).
func (s *Store) rateLimitClaimCooldown(ctx context.Context, subject string, nowUnix, cooldownS int64) (bool, error) {
	res, err := s.db.ExecContext(ctx,
		`INSERT INTO rate_limit_cooldowns (subject, last_open_at)
		 VALUES (?, ?)
		 ON CONFLICT (subject) DO UPDATE SET last_open_at = excluded.last_open_at
		 WHERE excluded.last_open_at - last_open_at >= ?`,
		subject, nowUnix, cooldownS)
	if err != nil {
		return false, err
	}
	n, err := res.RowsAffected()
	if err != nil {
		return false, err
	}
	return n > 0, nil
}

// NoteChatMessage counts an inbound chat message for a sender subject.
// quiet=true past the per-minute cap (the bot should stop replying, but the
// webhook still 200s). Fail-open on counter errors: keep replying.
func (s *Store) NoteChatMessage(ctx context.Context, envCfg RateLimitConfig, subject string, nowUnix int64) bool {
	cfg := s.ResolveRateLimitConfig(ctx, envCfg)
	ws := FixedWindowStart(nowUnix, MinuteS)
	if err := s.rateLimitBump(ctx, "chat_1m", subject, ws, 1); err != nil {
		return false
	}
	var count int64
	if err := s.db.QueryRowContext(ctx,
		`SELECT count FROM rate_limit_counters WHERE scope='chat_1m' AND subject=? AND window_start=?`,
		subject, ws).Scan(&count); err != nil {
		return false
	}
	return count > cfg.ChatMsgsPerMin
}
